cpp/src/arrow/filesystem/azurefs_test.cc

```diff
@@ -193,51 +265,123 @@ TEST(AzureFileSystem, OptionsCompare) {
+  std::string NotFoundObjectPath() const { return ContainerPath() + "not-found"; }
+
+  std::string RandomDirectoryPath(RNG& rng) {
+    return internal::ConcatAbstractPath(container_name, RandomDirectoryName(rng));
+  }
+
+  // Utilities
+
+  static std::string RandomContainerName(RNG& rng) { return RandomChars(32, rng); }
+  static std::string RandomDirectoryName(RNG& rng) { return RandomChars(32, rng); }
+
+  static std::string RandomLine(int lineno, std::size_t width, RNG& rng) {
+    auto line = std::to_string(lineno) + ":    ";
+    line += RandomChars(width - line.size() - 1, rng);
+    line += '\n';
+    return line;
+  }
+
+  static std::size_t RandomIndex(std::size_t end, RNG& rng) {
+    return std::uniform_int_distribution<std::size_t>(0, end - 1)(rng);
+  }
+
+  static std::string RandomChars(std::size_t count, RNG& rng) {
+    auto const fillers = std::string("abcdefghijlkmnopqrstuvwxyz0123456789");
+    std::uniform_int_distribution<std::size_t> d(0, fillers.size() - 1);
+    std::string s;
+    std::generate_n(std::back_inserter(s), count, [&] { return fillers[d(rng)]; });
+    return s;
+  }
+};
+
+class AzureFileSystemTest : public ::testing::Test {
+ protected:
+  // Set in constructor
+  std::mt19937_64 rng_;
+
+  // Set in SetUp()
+  int64_t debug_log_start_ = 0;
+  bool set_up_succeeded_ = false;
+  AzureOptions options_;
+
   std::shared_ptr<FileSystem> fs_;
   std::unique_ptr<Blobs::BlobServiceClient> blob_service_client_;
-  std::unique_ptr<Files::DataLake::DataLakeServiceClient> datalake_service_client_;
-  AzureOptions options_;
-  std::mt19937_64 generator_;
-  std::string container_name_;
-  bool suite_skipped_ = false;
+  std::unique_ptr<DataLake::DataLakeServiceClient> datalake_service_client_;
 
-  AzureFileSystemTest() : generator_(std::random_device()()) {}
+ public:
+  AzureFileSystemTest() : rng_(std::random_device()()) {}
+
+  virtual Result<BaseAzureEnv*> GetAzureEnv() const = 0;
 
-  virtual Result<AzureOptions> MakeOptions() = 0;
+  static Result<AzureOptions> MakeOptions(BaseAzureEnv* env) {
+    AzureOptions options;
+    options.backend = env->backend();
+    ARROW_EXPECT_OK(
+        options.ConfigureAccountKeyCredential(env->account_name(), env->account_key()));
+    return options;
+  }
 
   void SetUp() override {
-    auto options = MakeOptions();
-    if (options.ok()) {
-      options_ = *options;
+    auto make_options = [this]() -> Result<AzureOptions> {
+      ARROW_ASSIGN_OR_RAISE(auto env, GetAzureEnv());
+      EXPECT_THAT(env, NotNull());
+      ARROW_ASSIGN_OR_RAISE(debug_log_start_, env->GetDebugLogSize());
+      return MakeOptions(env);
+    };
+    auto options_res = make_options();
+    if (!options_res.ok() && options_res.status().IsCancelled()) {
+      GTEST_SKIP() << options_res.status().message();
     } else {
-      suite_skipped_ = true;
-      GTEST_SKIP() << options.status().message();
+      EXPECT_OK_AND_ASSIGN(options_, std::move(options_res));
```

Comment:
Do we need this explicit `std::move()`?

Reply:
Hmmm. I didn't know `EXPECT_OK_AND_ASSIGN` always moved. I thought it only moved when the `rexpr` was an `&&`.